File: frontend/src/infrastructure/repositories/interaction_repository.ts
/**
 * Repositorio de Interacciones.
 * Gestiona comentarios y visitas a ubicaciones.
 */

import api_client from '../api_client';
import type {
    Interaction_Model,
    Interaction_Create,
    Interaction_List_Response
} from '../../domain/models';

/**
 * Obtiene las interacciones de una ubicación.
 * @param location_id - ID de la ubicación.
 * @returns Promesa con la lista de interacciones.
 */
export const get_location_interactions = async (
    location_id: string
): Promise<Interaction_List_Response> => {
    const response = await api_client.get<Interaction_List_Response>(
        `/interactions/location/${location_id}`
    );
    return response.data;
};

/**
 * Crea una nueva interacción (comentario o visita).
 * @param data - Datos de la interacción a crear.
 * @returns Promesa con la interacción creada.
 */
export const create_interaction = async (
    data: Interaction_Create
): Promise<Interaction_Model> => {
    const response = await api_client.post<Interaction_Model>('/interactions', data);
    return response.data;
};

/**
 * Elimina una interacción.
 * @param interaction_id - ID de la interacción a eliminar.
 * @returns Promesa que se resuelve cuando se elimina.
 */
export const delete_interaction = async (interaction_id: string): Promise<void> => {
    await api_client.delete(`/interactions/${interaction_id}`);
};

File: backend/repositories/interaction_repository.py
"""
Repositorio para operaciones de Interacción en MongoDB.
"""

from datetime import datetime
from bson import ObjectId
from motor.motor_asyncio import AsyncIOMotorDatabase
from models.interaction import Interaction


class InteractionRepository:
    """
    Repositorio para gestionar interacciones en la base de datos.
    """
    
    def __init__(self, db: AsyncIOMotorDatabase):
        """
        Inicializa el repositorio con la conexión a la base de datos.
        :param db: Instancia de la base de datos MongoDB.
        """
        self.collection = db["interactions"]
    
    async def find_by_location(self, location_id: str) -> list[Interaction]:
        """
        Obtiene todas las interacciones de una ubicación.
        :param location_id: ID de la ubicación.
        :return: Lista de interacciones.
        """
        cursor = self.collection.find({"location_id": location_id}).sort("created_at", -1)
        interactions = []
        async for doc in cursor:
            doc["_id"] = str(doc["_id"])
            interactions.append(Interaction(**doc))
        return interactions
    
    async def find_by_user(self, user_email: str) -> list[Interaction]:
        """
        Obtiene todas las interacciones de un usuario.
        :param user_email: Email del usuario.
        :return: Lista de interacciones.
        """
        cursor = self.collection.find({"user_email": user_email}).sort("created_at", -1)
        interactions = []
        async for doc in cursor:
            doc["_id"] = str(doc["_id"])
            interactions.append(Interaction(**doc))
        return interactions
    
    async def create(self, interaction: Interaction) -> Interaction:
        """
        Crea una nueva interacción.
        :param interaction: Datos de la interacción a crear.
        :return: Interacción creada con su ID.
        """
        interaction_dict = interaction.model_dump(exclude={"id"})
        interaction_dict["created_at"] = datetime.utcnow()
        
        result = await self.collection.insert_one(interaction_dict)
        interaction_dict["_id"] = str(result.inserted_id)
        return Interaction(**interaction_dict)
    
    async def delete(self, interaction_id: str) -> bool:
        """
        Elimina una interacción.
        :param interaction_id: ID de la interacción a eliminar.
        :return: True si se eliminó, False si no existía.
        """
        try:
            result = await self.collection.delete_one({"_id": ObjectId(interaction_id)})
            return result.deleted_count > 0
        except Exception:
            return False
